htlcswitch/link.go

```diff
@@ -447,6 +453,22 @@ func (l *channelLink) Stop() {
 
 	close(l.quit)
 	l.wg.Wait()
+
+	// As a final precaution, we will attempt to flush any uncommitted
+	// preimages to the preimage cache. The preimages should be re-delivered
+	// after channel reestablishment, however this adds an extra layer of
+	// protection in case the peer never returns. Without this, we will be
+	// unable to settle any contracts depending on the preimages even though
+	// we had learned them at some point.
```

Comment:
Just to make sure that I understand this correctly: this is not to miss the opportunity to apply the preimages to the downstream node. But if this was not in, it would not lead to loss of funds would it? We would learn the preimage when the channel is force closed and then settle downstream?

Reply:
assuming the downstream node were to pull on chain, we may learn the preimage then. alternatively, if the peer reconnects and sends the same settle message and the commit sig, that would also suffice.

atm we spawn and forget a goroutine that is meant handle the write of each preimage, which in most cases would probably execute before shutdown/disconnect. however with the addition batching, we can't rely on these background tasks to execute for us, and in fact may not have attempted the write at all. so we add this to ensure that we always attempt to write any received preimages before the link exits, which is akin to what we do now, though the current impl has weaker consistency guarantees

Reply:
Ok clear. Yes that goroutine was not perfect indeed.